nsxt/resource_nsxt_upgrade_prepare.go

```diff
@@ -0,0 +1,451 @@
+	}
+	return resourceNsxtUpgradePrepareRead(d, m)
+}
+
+func resourceNsxtUpgradePrepareDelete(d *schema.ResourceData, m interface{}) error {
+	return nil
+}
+
+func uploadPrecheckAndUpgradeBundle(d *schema.ResourceData, m interface{}) error {
+	upgradeBundleURL := d.Get("upgrade_bundle_url").(string)
+	precheckBundleURL := d.Get("precheck_bundle_url").(string)
+	upgradeBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_UPGRADE
+	precheckBundleType := nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_PRE_UPGRADE
+	timeout := d.Get("bundle_upload_timeout").(int)
+	c := m.(nsxtClients)
+	userName := c.NsxtClientConfig.UserName
+	password := c.NsxtClientConfig.Password
+	if !precheckBundleCompatibilityCheck(precheckBundleURL) {
+		return fmt.Errorf("Precheck bundle is only supported and is required for NSXT version >= 4.1.1")
+	}
+	if len(precheckBundleURL) > 0 {
+		err := uploadUpgradeBundle(m, precheckBundleURL, userName, password, precheckBundleType, timeout)
+		if err != nil {
+			return fmt.Errorf("Failed to upload precheck bundle: %s", err)
+		}
+	}
+	err := uploadUpgradeBundle(m, upgradeBundleURL, userName, password, upgradeBundleType, timeout)
+	if err != nil {
+		return fmt.Errorf("Failed to upload upgrade bundle: %s", err)
+	}
+	return nil
+}
+
+func precheckBundleCompatibilityCheck(precheckBundleURL string) bool {
+	if nsxVersionLower("4.1.1") && len(precheckBundleURL) > 0 {
+		return false
+	}
+	if nsxVersionHigherOrEqual("4.1.1") && len(precheckBundleURL) == 0 {
+		return false
+	}
+	return true
+}
+
+func uploadUpgradeBundle(m interface{}, url string, userName string, password string, bundleType string, timeout int) error {
+	connector := getPolicyConnector(m)
+	summaryClient := upgrade.NewSummaryClient(connector)
+	summary, err := summaryClient.Get()
+	if err != nil {
+		return err
+	}
+	if bundleType == nsxModel.UpgradeBundleFetchRequest_BUNDLE_TYPE_UPGRADE {
+		// Check if same upgrade bundle has already been uploaded, if so then skip uploading.
+		// example bundle name from NSX API: VMware-NSX-upgrade-bundle-5.0.0.0.0.23236379.mub
+		if summary.UpgradeBundleFileName != nil {
+			existingBundleName := *summary.UpgradeBundleFileName
+			arr := strings.Split(url, "/")
+			bundleName := arr[len(arr)-1]
+			if existingBundleName == bundleName {
+				log.Printf("Upgrade bundle already uploaded to NSX, skip uploading upgrade bundle")
+				return nil
+			}
+		}
+	} else {
+		// Check if same precheck bundle has already been uploaded, if so then skip uploading.
+		// example bundle version from NSX API: 5.0.0.0.0.23236379
+		// example precheck bundle name: VMware-NSX-upgrade-bundle-5.0.0.0.0.23236379-pre-check.pub
+		if summary.PreUpgradeBundleVersion != nil {
+			existingBundleVersion := *summary.PreUpgradeBundleVersion
+			arr := strings.Split(url, "/")
+			bundleName := arr[len(arr)-1]
+			bundleVersionArr := strings.Split(bundleName, "-")
+			if len(bundleVersionArr) != 7 {
```

Comment:
I'm not sure we should rely on the name from url, since it could be anything..

Reply:
Unfortunately, it seems the only way to get uploaded bundle information is through the upgrade/summary API, which will return the name of uploaded mub bundle and the version of uploaded pub bundle... If user decides to rename the bundle to a random format I'm afraid we cannot check if the same bundle has been uploaded

Reply:
Added `version` back to the schema of the resource, it will be used to determine if precheck bundle needs re-upload. For upgrade bundle, we use d.HasChange(bundleURL) to decide if reload is needed.